### ISSUE TITLE
Increase test coverage

### ISSUE THREAD

**Issue by tomrijnbeek** [beardgame/utilities#234]
Any new PR should be accompanied by relevant tests. However, large portions of the libraries are as of yet untested. Any PRs improving test coverage of existing code are welcome.

Keep the following things in mind when creating tests:

* Follow existing practices to keep the style of tests consistent. [This](https://github.com/beardgame/utilities/blob/master/Bearded.Utilities.Tests/Geometry/CircularArc2Tests.cs) is a good example to start from.
* Consider using property tests wherever possible, as they ensure our code does not only work for a set of hand-picked values.
* Focus on testing behaviours rather than specific members.
* Refactoring code as part of adding tests is allowed and encouraged, though consider verifying breaking changes before making them.
* Consider creating a draft PR before starting to indicate you are working on the tests for a specific class. Likewise, check if there aren't any open PRs for the class you want to test ahead of starting work.

If you are unsure what class to test, but are interested to help, feel free to reply below and we're happy to get you going.

**Comment by amulware** [comment]
See below a list of all namespaces and types and their coverage (at time of writing, but let's see if we can keep it updated), in case you don't have your own coverage tool.
<details>
<summary>🚧 22% Bearded.Utilities</summary>

- [ ] 0% AsyncAtomicUpdating&lt;T&gt;
- [ ] 0% Box&lt;T&gt;
- [ ] 0% Do
- [ ] 0% Environment
- [ ] 0% Id&lt;T&gt;
- [ ] 0% IdManager
- [ ] 0% Interpolate
- [ ] 2% MathExtensions
- [ ] 0% Mathf
- [ ] 94% Maybe&lt;T&gt;
- [x] 100% Maybe
- [ ] 0% MoreMath
- [x] 100% RandomExtensions
- [ ] 0% ResettableLazy
- [ ] 0% ResettableLazy&lt;T&gt;
- [ ] 0% Singleton&lt;TSelf&gt;
- [x] 100% StaticRandom
- [ ] 0% Void

</details>
<details>
<summary>🚧 89% Bearded.Utilities.Algorithms</summary>

- [ ] 84% BinPacking
  - [x] 100% PositionedRectangle&lt;T&gt;
  - [x] 100% Rectangle&lt;T&gt;
  - [x] 100% Result&lt;T&gt;
  - [ ] 95% Tree&lt;T&gt;
    - [x] 100% Node
- [ ] 95% CoffmanGraham
  - [x] 100% ArbitraryGraphSolver
  - [ ] 83% DecreasingNumberSequence
  - [x] 100% ReducedGraphSolver
- [ ] 91% HungarianAlgorithm

</details>
<details>
<summary>🚧 62% Bearded.Utilities.Collections</summary>

- [ ] 98% DeletableObjectList&lt;T&gt;
- [ ] 97% DeletableObjectListEnumerator&lt;T&gt;
- [ ] 0% IdDictionary&lt;T&gt;
- [ ] 0% MutableLinkedList&lt;T&gt;
- [ ] 0% MutableLinkedListEnumerator&lt;T&gt;
- [ ] 0% MutableLinkedListNode
- [ ] 0% MutableLinkedListNode&lt;T&gt;
- [ ] 80% PrefixTrie
  - [x] 100% Node
- [ ] 91% PriorityQueue&lt;TPriority,TValue&gt;
- [ ] 99% StaticPriorityQueue&lt;TPriority,TValue&gt;

</details>
<details>
<summary>🚧 21% Bearded.Utilities.Geometry</summary>

- [ ] 18% Angle
- [ ] 0% Arc&lt;T&gt;
- [ ] 0% Arc2
- [ ] 0% Arc3
- [ ] 0% Bezier2nd2
- [ ] 0% Bezier2nd3
- [ ] 0% Bezier3rd2
- [ ] 0% Bezier3rd3
- [ ] 88% CircularArc2
- [ ] 39% Direction2
- [ ] 33% GeometryExtensions
- [ ] 0% Interval
- [ ] 0% PolarPosition
- [ ] 0% Rectangle

</details>

⬤ ✅ 100% Bearded.Utilities.Graphs

<details>
<summary>⛔ 0% Bearded.Utilities.Input</summary>

- [ ] 0% GamePadStateManager
- [ ] 0% InputAction
  - [ ] 0% AnyAction
  - [ ] 0% BinaryAction
  - [ ] 0% OrAction
- [ ] 0% InputManager
  - [ ] 0% ActionConstructor
  - [ ] 0% KeyboardActions
  - [ ] 0% MouseActions
- [ ] 0% KeyboardEvents

</details>
<details>
<summary>⛔ 0% Bearded.Utilities.Input.Actions</summary>

- [ ] 0% DeferredAction
  - [ ] 0% LambdaAction
- [ ] 0% DigitalAction
- [ ] 0% DummyAction
- [ ] 0% KeyboardAction
- [ ] 0% MouseButtonAction

</details>
<details>
<summary>⛔ 0% Bearded.Utilities.IO</summary>

- [ ] 0% FileModifiedWatcher
- [ ] 0% Logger
  - [ ] 0% Entry
  - [ ] 0% Writer

</details>
<details>
<summary>🚧 32% Bearded.Utilities.Linq</summary>

- [ ] 32% Extensions

</details>
<details>
<summary>🚧 85% Bearded.Utilities.Monads</summary>

- [x] 100% Failure&lt;T&gt;
- [ ] 89% Result&lt;TResult,TError&gt;
- [ ] 50% Result
- [x] 100% Success&lt;T&gt;

</details>
<details>
<summary>🚧 8% Bearded.Utilities.SpaceTime</summary>

- [ ] 16% Acceleration
- [ ] 10% Acceleration2
- [ ] 0% Acceleration3
- [ ] 0% AddDimensionExtensions
- [ ] 0% AngularAcceleration
- [ ] 0% AngularVelocity
- [ ] 20% Difference2
- [ ] 0% Difference3
- [ ] 17% Extensions
- [ ] 0% Frequency
- [x] 100% GravitationalConstant
- [ ] 0% Instant
- [ ] 14% Mass
- [ ] 0% Position2
- [ ] 0% Position3
- [ ] 0% Speed
- [ ] 23% Squared&lt;T&gt;
- [ ] 0% TimeSpan
- [ ] 13% Unit
- [ ] 0% Velocity2
- [ ] 0% Velocity3

</details>
<details>
<summary>⛔ 0% Bearded.Utilities.Threading</summary>

- [ ] 0% BackgroundActionQueue
- [ ] 0% ManualActionQueue

</details>
<details>
<summary>⛔ 0% Bearded.Utilities.Tilemaps</summary>

- [ ] 0% Step

</details>
<details>
<summary>🚧 32% Bearded.Utilities.Tilemaps.Rectangular</summary>

- [ ] 0% Extensions
- [ ] 34% Tile&lt;TTileValue&gt;
- [ ] 68% Tilemap&lt;TTileValue&gt;

</details>
<details>
<summary>🚧 22% Bearded.Utilities</summary>

- [ ] 0% AsyncAtomicUpdating&lt;T&gt;
- [ ] 0% Box&lt;T&gt;
- [ ] 0% Do
- [ ] 0% Environment
- [ ] 0% Id&lt;T&gt;
- [ ] 0% IdManager
- [ ] 0% Interpolate
- [ ] 2% MathExtensions
- [ ] 0% Mathf
- [ ] 94% Maybe&lt;T&gt;
- [x] 100% Maybe
- [ ] 0% MoreMath
- [x] 100% RandomExtensions
- [ ] 0% ResettableLazy
- [ ] 0% ResettableLazy&lt;T&gt;
- [ ] 0% Singleton&lt;TSelf&gt;
- [x] 100% StaticRandom
- [ ] 0% Void

</details>
<details>
<summary>🚧 89% Bearded.Utilities.Algorithms</summary>

- [ ] 84% BinPacking
  - [x] 100% PositionedRectangle&lt;T&gt;
  - [x] 100% Rectangle&lt;T&gt;
  - [x] 100% Result&lt;T&gt;
  - [ ] 95% Tree&lt;T&gt;
    - [x] 100% Node
- [ ] 95% CoffmanGraham
  - [x] 100% ArbitraryGraphSolver
  - [ ] 83% DecreasingNumberSequence
  - [x] 100% ReducedGraphSolver
- [ ] 91% HungarianAlgorithm

</details>
<details>
<summary>🚧 62% Bearded.Utilities.Collections</summary>

- [ ] 98% DeletableObjectList&lt;T&gt;
- [ ] 97% DeletableObjectListEnumerator&lt;T&gt;
- [ ] 0% IdDictionary&lt;T&gt;
- [ ] 0% MutableLinkedList&lt;T&gt;
- [ ] 0% MutableLinkedListEnumerator&lt;T&gt;
- [ ] 0% MutableLinkedListNode
- [ ] 0% MutableLinkedListNode&lt;T&gt;
- [ ] 80% PrefixTrie
  - [x] 100% Node
- [ ] 91% PriorityQueue&lt;TPriority,TValue&gt;
- [ ] 99% StaticPriorityQueue&lt;TPriority,TValue&gt;

</details>
<details>
<summary>🚧 21% Bearded.Utilities.Geometry</summary>

- [ ] 18% Angle
- [ ] 0% Arc&lt;T&gt;
- [ ] 0% Arc2
- [ ] 0% Arc3
- [ ] 0% Bezier2nd2
- [ ] 0% Bezier2nd3
- [ ] 0% Bezier3rd2
- [ ] 0% Bezier3rd3
- [ ] 88% CircularArc2
- [ ] 39% Direction2
- [ ] 33% GeometryExtensions
- [ ] 0% Interval
- [ ] 0% PolarPosition
- [ ] 0% Rectangle

</details>

⬤ ✅ 100% Bearded.Utilities.Graphs

<details>
<summary>⛔ 0% Bearded.Utilities.Input</summary>

- [ ] 0% GamePadStateManager
- [ ] 0% InputAction
  - [ ] 0% AnyAction
  - [ ] 0% BinaryAction
  - [ ] 0% OrAction
- [ ] 0% InputManager
  - [ ] 0% ActionConstructor
  - [ ] 0% KeyboardActions
  - [ ] 0% MouseActions
- [ ] 0% KeyboardEvents

</details>
<details>
<summary>⛔ 0% Bearded.Utilities.Input.Actions</summary>

- [ ] 0% DeferredAction
  - [ ] 0% LambdaAction
- [ ] 0% DigitalAction
- [ ] 0% DummyAction
- [ ] 0% KeyboardAction
- [ ] 0% MouseButtonAction

</details>
<details>
<summary>⛔ 0% Bearded.Utilities.IO</summary>

- [ ] 0% FileModifiedWatcher
- [ ] 0% Logger
  - [ ] 0% Entry
  - [ ] 0% Writer

</details>
<details>
<summary>🚧 32% Bearded.Utilities.Linq</summary>

- [ ] 32% Extensions

</details>
<details>
<summary>🚧 85% Bearded.Utilities.Monads</summary>

- [x] 100% Failure&lt;T&gt;
- [ ] 89% Result&lt;TResult,TError&gt;
- [ ] 50% Result
- [x] 100% Success&lt;T&gt;

</details>
<details>
<summary>🚧 8% Bearded.Utilities.SpaceTime</summary>

- [ ] 16% Acceleration
- [ ] 10% Acceleration2
- [ ] 0% Acceleration3
- [ ] 0% AddDimensionExtensions
- [ ] 0% AngularAcceleration
- [ ] 0% AngularVelocity
- [ ] 20% Difference2
- [ ] 0% Difference3
- [ ] 17% Extensions
- [ ] 0% Frequency
- [x] 100% GravitationalConstant
- [ ] 0% Instant
- [ ] 14% Mass
- [ ] 0% Position2
- [ ] 0% Position3
- [ ] 0% Speed
- [ ] 23% Squared&lt;T&gt;
- [ ] 0% TimeSpan
- [ ] 13% Unit
- [ ] 0% Velocity2
- [ ] 0% Velocity3

</details>
<details>
<summary>⛔ 0% Bearded.Utilities.Threading</summary>

- [ ] 0% BackgroundActionQueue
- [ ] 0% ManualActionQueue

</details>
<details>
<summary>⛔ 0% Bearded.Utilities.Tilemaps</summary>

- [ ] 0% Step

</details>
<details>
<summary>🚧 32% Bearded.Utilities.Tilemaps.Rectangular</summary>

- [ ] 0% Extensions
- [ ] 34% Tile&lt;TTileValue&gt;
- [ ] 68% Tilemap&lt;TTileValue&gt;

</details>

**Comment by MaurizioPz** [comment]
I would like to do part of this issue, maybe starting with Bearded.Utilities.Linq. Is it ok?

**Comment by amulware** [comment]
@MaurizioPz Hey, definitely! Feel free to submit a PR for a class at a time, or in the case of the Linq extensions perhaps even per method, if it's one of the more complex ones. :)
